### PR TITLE
ENH: Multiple comment tokens in loadtxt

### LINKED ISSUE #2633: support multiple comment strings in loadtxt() (Trac #2041)
_Original ticket http://projects.scipy.org/numpy/ticket/2041 on 2012-02-08 by trac user andreash, assigned to unknown._

I would like to see the following enhancement in !NumPy:

   In the numpy.loadtxt method, support passing multiple comment strings, perhaps as a list.

In reality, people change dataformats quite often, and too many people don't care too much about the users of their data (at least in my field of science). So when processing large datasets, I have to deal with files which either have `#` or `*` as comment string. If !NumPy supported something like the following, I'd be happy:

    loadtxt('myfile.txt', comments=['*','#'])

### PR DESCRIPTION
**Pull request by Nodd**
Allows to use multiple comment tokens instead of a unique string.

Closes gh-2633

### CODE REVIEW

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -715,13 +716,20 @@ def loadtxt(fname, dtype=float, comments='#', delimiter=None,
...
+        comments = [asbytes(comment) for comment in comments]
     user_converters = converters
     if delimiter is not None:
         delimiter = asbytes(delimiter)
     if usecols is not None:
         usecols = list(usecols)
 
+    # Compile regex for comments beforehand
+    regex_comments = re.compile(asbytes('|').join(re.escape(comment)
+                                                  for comment in comments))
```
Comment (charris) [inline numpy/lib/npyio.py]: Needs PEP8 alignment.

Comment (charris) [inline numpy/lib/npyio.py]: NVM.

Comment (charris) [inline numpy/lib/npyio.py]: Not as easy to read as it could be. Might be better to separate the regex string definition so that it fits on one line.

**Comment by charris** [comment]
LGTM. @njsmith Look good to you?

**Comment by Nodd** [comment]
I resolved the conflict and took care of the long line.

**Comment by firegurafiku** [comment]
At my university I have to process a lot of files produced by Agilent's oscilloscopes which use both "#" and "!" comments at the same time. So, I'd like to see this PR merged. Can I do something for this to happen?

**Comment by Nodd** [comment]
I rebased it (again).

**Comment by Nodd** [comment]
So is there anything missing that prevents merging this PR ? Main developer time maybe ? ;)
I started it one year ago, rebased twice, and at least 2 users want this functionality...

**Comment by firegurafiku** [comment]
@njsmith, @rgommers, anybody? We need a bit of your attention.

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -791,7 +792,14 @@ def loadtxt(fname, dtype=float, comments='#', delimiter=None,
     """
     # Type conversions for Py3 convenience
     if comments is not None:
-        comments = asbytes(comments)
+        if isinstance(comments, (np.compat.basestring, np.compat.bytes)):
```
Comment (stefanv) [inline numpy/lib/npyio.py]: The PR looks reasonable to me; this is the only "paragraph" I'm not sure about

Comment (Nodd) [inline numpy/lib/npyio.py]: You mean I should check against `basestring` only ? It could potentially create a compatibility issue.

Comment (stefanv) [inline numpy/lib/npyio.py]: I simply meant that I am not well versed in numpy's compatibility layer, so I'd like someone else on the team to have a look.

Comment (charris) [inline numpy/lib/npyio.py]: This is fine. Howeve, both `basestring` and `bytes` are already imported, so the `np.compat` prefix isn't needed.

**Comment by Nodd** [comment]
I removed the unneeded `np.compat` prefix.

**Comment by stefanv** [comment]
@charris @njsmith Shall we get this one merged before another rebase is required?

**Comment by charris** [comment]
@Nodd Thanks, sorry about the delay.

**Pull request merged**